### PR TITLE
[A1] Broken Access Control - Vulnerable Ecommerce API

### PR DESCRIPTION
**Pull request by henriporto**
## This solution refers to which of the apps?

A1 - Broken Access Control | Golang | Vulnerable Ecommerce API

## What did you do to mitigate the vulnerability?

I implemented a token check before sending the ticket.

## Did you test your changes? What commands did you run?

Yes. I simulated the attack with Burp Suite, changing the route.

### CODE REVIEW

File: owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go [hunk 1]
```diff
@@ -16,22 +19,43 @@ func HealthCheck(c echo.Context) error {
...
+		if err == jwt.ErrSignatureInvalid {
+			return c.JSON(http.StatusUnauthorized, map[string]string{"result": "error", "details": "Error StatusUnauthorized."})
+		}
+		return c.JSON(http.StatusBadRequest, map[string]string{"result": "error", "details": "Error StatusBadRequest."})
 	}
+	if !tkn.Valid {
+		return c.JSON(http.StatusUnauthorized, map[string]string{"result": "error", "details": "Error StatusUnauthorized."})
+	}
+
+	if claims.Username == userDataResult.Username{
```
Comment (gustavocovas) [inline owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go:47]: Now that we are getting the username from the JWT, we could remove the username parameter and avoid this check. What do you think?

File: owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go [hunk 2]
```diff
@@ -16,22 +19,43 @@ func HealthCheck(c echo.Context) error {
...
+				"ticket":   userDataResult.Ticket,
+			})
+		}
 
-	msgTicket := fmt.Sprintf("Hey, %s! This is your ticket: %s\n", userDataResult.Username, userDataResult.Ticket)
-	return c.String(http.StatusOK, msgTicket)
+		msgTicket := fmt.Sprintf("Hey, %s! This is your ticket: %s\n", userDataResult.Username, userDataResult.Ticket)
+		return c.String(http.StatusOK, msgTicket)
+	}
+	return c.JSON(http.StatusBadRequest, map[string]string{"result": "error", "details": "Access denied."})
```
Comment (gustavocovas) [inline owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go:60]: Before this last validation, all checks are something like:
```
if something is wrong {
  return error
}
```
However, here this logic is inverted (we return success inside the if clause). Could you fix this to make the code more consistent?

**Review by gustavocovas — COMMENTED** [review]
Very nice! Using JWT is a battle-tested way of implement access control. I see that you also removed some hard-coded credentials from the source code 🐼 
I just left some suggestions to make the app more simple and concise, could you take a look at them?

File: owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go (hunk 2 above)
Comment (henriporto) [reply owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go]: Thanks for the sugestion. Done!

File: owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go (hunk 1 above)
Comment (gustavocovas) [reply owasp-top10-2021-apps/a1/ecommerce-api/app/handlers/handlers.go]: Cool, but we are still reading the user id from the parameters and doing the query afterwards:

```go
id := c.Param("id")
userDataQuery := map[string]interface{}{"userID": id}
```

You could do something such as:

```go
// Parse JWT claims
userID, _ := claims["id"]
userDataQuery := map[string]interface{}{"userID": userID}
```

If you want to try that, take a look at [this PR](https://github.com/globocom/secDevLabs/pull/275).

**Comment by gustavocovas** [comment]
Hey, I just left some comments about the `user_id`. As we discussed, you fixed the broken access control vulnerability, nice job! Feel free to re-open if you want to try to remove the `user_id` parameter and avoid performing queries that would result in an `access denied` response.

**Comment by gustavocovas** [comment]
Great! Nice job 🐼

**Pull request closed without merging**